### DISCUSSION TITLE
Luarocks fails to resolve manifest

### DISCUSSION THREAD

**Discussion by adwaymalhotra** [lumen-oss/rocks.nvim#329] [Q&A]
I'm facing an issue today where luarocks just can't seem to find some plugins. I keep seeing these `temporary failure in name resolution` errors. 
```
ERROR | 2024-05-13 17:24:46 | ...re/nvim/rocks/rocks_rtp/lua/rocks/operations/helpers.lua:68 | Failed to install mini.clue
ERROR | 2024-05-13 17:24:46 | ....local/share/nvim/rocks/rocks_rtp/lua/rocks/luarocks.lua:66 | luarocks CLI FAILED
ERROR | 2024-05-13 17:24:46 | ....local/share/nvim/rocks/rocks_rtp/lua/rocks/luarocks.lua:67 | Warning: Failed searching manifest: Failed downloading https://nvim-neorocks.github.io/rocks-binaries/dev/manifest-5.1 - HTTP/1.1 404 Not Found
Warning: Failed searching manifest: Failed downloading https://luarocks.org/dev/manifest-5.1 - temporary failure in name resolution
Warning: Failed searching manifest: Failed downloading https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/dev/manifest-5.1 - HTTP/1.1 404 Not Found
Warning: Failed searching manifest: Failed downloading https://loadk.com/luarocks/dev/manifest-5.1 - HTTP/1.1 404 Not Found
Warning: Failed searching manifest: Failed downloading https://nvim-neorocks.github.io/rocks-binaries/manifest-5.1 - temporary failure in name resolution
Warning: Failed searching manifest: Failed downloading https://luarocks.org/manifest-5.1 - temporary failure in name resolution
Warning: Failed searching manifest: Failed downloading https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/manifest-5.1 - temporary failure in name resolution
Warning: Failed searching manifest: Failed downloading https://loadk.com/luarocks/manifest-5.1 - temporary failure in name resolution

Error: No results matching query were found for Lua 5.1.
To check if it is available for other Lua versions, use --check-lua-versions.
```

Curiously, when I use the luarocks from the one that rocks.nvim installer has set up (version name dev), then it doesn't look at `https://luafr.org`. When I use the luarocks packaged in debian on the command, it fails to resolve the urls above, but it does look at `luafr` and resolves some of the manifests successfully. That's how I managed to install rocks.nvim.

The issue seems to be localised to luarocks. The URLs are reachable on all browsers, through ping and wget/curl in the terminal. 

Anyone got a clue?

Edit: I should add that in my infinite wisdom, I decided to rm my .local/share/nvim directory. So, atm I'm stuck without any plugins unless I install each one-by-one using luarocks on the commandline (instead of through rocks). I really don't want to move back to lazy... :(

**Comment by mrcjkb** [comment]
Hey :wave: 

From the logs, it looks like it's searching for dev manifests (see the `/dev/` route in the URIs), which is what it does when you try to install a `dev` version (rocks.nvim will prompt you to do so, if it can't find a rock, so I'm presuming you confirmed?).

Most luarocks servers (including `rocks-binaries` and `loadk.com`) don't have a `dev` manifest, which would explain the 404 warnings (I wouldn't worry about them).

The `temporary failure in name resolution` warnings indeed look like something localised to luarocks.

The reason luarocks doesn't look at luafr.org when used by rocks.nvim, is likely because we pass in a luarocks config, so it will ignore any home or system luarocks config.
You should see a `luarocks-config.lua` file in your `~/.local/share/nvim/rocks` directory.
As a workaround, you could try adding the server there. If that works, we can create an issue for making it possible to override that config.

**Reply by adwaymalhotra** [reply]
Thanks for the response!

These logs in particular occur when trying to do a `Rocks sync` so some plugins have a dev version only - mind this is a setup that already worked for me. `Rocks update` failed in a similar way so I decided to nuke the whole thing.

I'm not sure how to do add servers to the luarocks-config.lua. This is what it looks like now after I added a rocks_servers array:
```
lua_version = "5.1"
rocks_trees = {
    {
      name = "rocks.nvim",
      root = "/home/adway/.local/share/nvim/rocks",
    },
}
rocks_servers = { "https://luafr.org/luarocks/" }
```
This does not seem to have an effect on either running `Rocks sync` in nvim or running `~/.local/share/nvim/luarocks ...` in the terminal.

**Reply by mrcjkb** [reply]
Hmm, that config looks correct to me.
Do the logs look the same? 

> or running ~/.local/share/nvim/luarocks ... in the terminal.

This is to be expected. rocks.nvim should pass the config by passing an `LUAROCKS_CONFIG` environment variable it won't be picked up when invoking it from the cli.

🤔 Maybe there's a bug in rocks.nvim and the config isn't being picked up properly? 
I'll have to test that later when I'm near my laptop.

**Reply by mrcjkb** [reply]
Ah, I think I know why it doesn't take effect.
rocks.nvim currently resets the `luarocks-config.lua` file to the default content at the beginning of every session.

**Reply by mrcjkb** [reply]
https://github.com/nvim-neorocks/rocks.nvim/issues/330

**Reply by adwaymalhotra** [reply]
Thanks for looking into it! 

I'm convinced it's a bug in luarocks somehow, because those urls (the ones which cause a name resolution error) work fine when I use curl/wget and in the browser, and I get the same error when I run the luarocks commands outside rocks.nvim. The difference in the servers is probably because of the difference in the versions of the luarocks binary. The dev version seems to not have `luafr.org` in the list. 

For the moment, I've gone back to lazy, but I'll try to get things working with rocks again on the weekend.

This happens to me on multiple systems (though both are corp machines, so may be that's why nobody else seems to face this issue).

**Reply by mrcjkb** [reply]
@adwaymalhotra you can now override the luarocks config by setting `vim.g.rocks_nvim.luarocks_config = "<path-to-file>"`.

There was an error in the snippet I posted earlier, by the way. It should be `lua_version = "5.1"`, not `lua_version = 5.1`.

This snippet should be correct:

```lua
lua_version = "5.1"
rocks_trees = {
    {
      name = "rocks.nvim",
      root = "/home/adway/.local/share/nvim/rocks",
    },
}
rocks_servers = { "https://luafr.org/luarocks/" }
```
